executor/utils.go

```diff
@@ -53,3 +58,201 @@ func deleteFromSet(set []string, value string) []string {
+	// to avoid missing information when statements_summary is just cleared.
+	stmt := "select digest, digest_text from information_schema.statements_summary union distinct " +
+		"select digest, digest_text from information_schema.statements_summary_history"
+	if queryGlobal {
+		stmt = "select digest, digest_text from information_schema.cluster_statements_summary union distinct " +
+			"select digest, digest_text from information_schema.cluster_statements_summary_history"
+	}
+	// Add the where clause if `inValues` is specified.
+	if len(inValues) > 0 {
+		stmt += " where digest in (" + strings.Repeat("%?,", len(inValues)-1) + "%?)"
+	}
+
+	stmtNode, err := exec.ParseWithParams(ctx, stmt, inValues...)
+	if err != nil {
+		return nil, err
+	}
+	rows, _, err := exec.ExecRestrictedStmt(ctx, stmtNode)
+	if err != nil {
+		return nil, err
+	}
+
+	res := make(map[string]string, len(rows))
+	for _, row := range rows {
+		res[row.GetString(0)] = row.GetString(1)
+	}
+	return res, nil
+}
+
+func (r *SQLDigestTextRetriever) updateWithQueryResult(queryResult map[string]string) {
+	for digest, text := range r.SQLDigestsMap {
+		if len(text) > 0 {
+			// The text of this digest is already known
+			continue
+		}
+		sqlText, ok := queryResult[digest]
+		if ok {
+			r.SQLDigestsMap[digest] = sqlText
+		}
+	}
+}
+
+// RetrieveLocal tries to retrieve the SQL text of the SQL digests from local information.
+func (r *SQLDigestTextRetriever) RetrieveLocal(ctx context.Context, sctx sessionctx.Context) error {
+	if len(r.SQLDigestsMap) == 0 {
+		return nil
+	}
+
+	var queryResult map[string]string
+	if len(r.SQLDigestsMap) <= r.fetchAllLimit {
+		inValues := make([]interface{}, 0, len(r.SQLDigestsMap))
+		for key := range r.SQLDigestsMap {
+			inValues = append(inValues, key)
+		}
+		var err error
+		queryResult, err = r.runQuery(ctx, sctx, false, inValues)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		if len(queryResult) == len(r.SQLDigestsMap) {
+			r.SQLDigestsMap = queryResult
+			return nil
+		}
+	} else {
+		var err error
+		queryResult, err = r.runQuery(ctx, sctx, false, nil)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	r.updateWithQueryResult(queryResult)
+	return nil
+}
+
+// RetrieveGlobal tries to retrieve the SQL text of the SQL digests from the information of the whole cluster.
+func (r *SQLDigestTextRetriever) RetrieveGlobal(ctx context.Context, sctx sessionctx.Context) error {
+	err := r.RetrieveLocal(ctx, sctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	var unknownDigests []interface{}
+	for k, v := range r.SQLDigestsMap {
+		if len(v) == 0 {
+			unknownDigests = append(unknownDigests, k)
+		}
+	}
+
+	if len(unknownDigests) == 0 {
+		return nil
+	}
+
+	var queryResult map[string]string
+	if len(r.SQLDigestsMap) <= r.fetchAllLimit {
+		queryResult, err = r.runQuery(ctx, sctx, true, unknownDigests)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	} else {
+		queryResult, err = r.runQuery(ctx, sctx, true, nil)
```

Comment:
Seems no result number limit here? Could it consumes much memory?

Reply:
I asked @crazycs520 before and he told me the statements_summary_history is unlikely to cause OOM in real situations, but I'm afraid so. Currently I didn't limit it for simplicity. Do you have any idea about this? Should we always use the `WHERE  IN` clause to filter the result?

Reply:
Theoretically it's better to use limit and where clause in queries,  if the history results will not be too big I think it's ok to keep the implementation here to make it simple.